A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+A48: xDS Least Request LB Policy
+----
+* Author(s): erikjoh & [fabric](https://github.com/orgs/spotify/teams/fabric)
+* Approver: markdroth
+* Status: Implementation in progress
+* Implemented in: Java in progress
+* Last updated: 2021-11-02
+* Discussion at: https://groups.google.com/g/grpc-io/c/4qycdcFfMUs
+
+## Abstract
+
+Add support for least request load balancing configured via xDS.
+
+## Background
+
+In multi-tenancy environments like for example Kubernetes, some upstream endpoints may perform much slower than others.
+When using plain round-robin load balancing, this becomes a problem since all endpoints get an equal amount of RPCs
+even when some endpoints perform worse. The result is increased tail latencies and in extreme cases even increased error rates.
+As an alternative to round-robin, Envoy provides a "least request"
+[load balancing policy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request)
+which uses per-endpoint outstanding/active request counters as a heuristic for determining load on each endpoint.
+The Envoy least request implementation uses two different algorithms depending on whether endpoints have non-equal
+weights or not. In the "*all weights equal*" case, Envoy samples N random available hosts and selects the one with
+fewest outstanding requests. In the "*all weights not equal*" case, Envoy switches to a weighted round robin schedule
+where weights are dynamically adjusted based on the outstanding requests.
+
+### Related Proposals: 
+
+This proposal builds on earlier work described in the following gRFCs:
+* [gRFC A27: xDS-based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+* [gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+Support `LEAST_REQUEST` LB policy and configuration of it using xDS.
+Specifically, gRPC will only support the "*all weights equal*" least request algorithm as
```

Comment:
How will the policy react if it gets more than one instance of the same address in the address list?  That's not likely to happen in the xDS case, but it can happen with DNS or other custom resolvers.

In this case, will the policy de-dup the addresses and assume just one instance, or will it treat each instance of the address as a separate address to be chosen from?

Reply:
Hmmm, interesting edge-case!
First of all, I assume you mean ip+port by "address" in this case (so e.g. "10.1.2.3:443")?
If a dns/custom resolver today receives duplicate addresses I presume it's being done to more heavily weigh traffic to some address when using round_robin.
If least_request were to de-dup in such a case, it would probably defeat the purpose behind having duplicate addresses in the first place.
And while the traffic weigh would not work nearly the same between RR and LR in this case, there would still be a somewhat heavier preference towards addresses that are duplicated in LR.
So if round_robin today doesn't do any de-dup of addresses and has a 1:1 mapping between addresses and subchannels I think least_request should probably follow that approach.

Does that make sense?
If so, then I can add this clarification in here somewhere :smile:

Reply:
Added `Allow picking same candidate multiple times` section further down in the gRFC.

Reply:
Hmm, I'm now thinking that maybe we went the wrong way with this, for two reasons.

First, it seems like the idea here is to have a different counter for each instance of the address.  However, given the discussion we just finished about where to store the counters, it seems a little strange to store two different counters for the same subchannel.

Second, in principle, having the same address listed multiple times is just another way of handling weighting, so I'm thinking that we should just consider this another way of expressing per-endpoint weight.  In other words, the equally weighted address list [A, B, A] should be considered equivalent to the weighted address list [A=2, B=1].  That's basically what the ring_hash policy does, so it seems consistent to do the same thing here.  And then the way we handle this becomes just part of the question of whether we support the "all requests not equal" algorithm.

WDYT?

Reply:
> it seems a little strange to store two different counters for the same subchannel.

I can't speak for C-core or Go until checking the implementations more carefully, but in Java it looks like duplicate addresses in the resolution result ([here](https://github.com/grpc/grpc-java/blob/a2398ce5dbe2eacf841d7af9c9aa5e912e8db0aa/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java#L73)) also causes duplicate/separate subchannels. In this case, I think it still makes sense to have separate counters on these subchannels as it would increase the selection probability (even though it's not proportional to the "weight").
This would also make the implementation simpler, since more logic from round robin may be used as-is.

If we end up adding the Weighted LR algorithm, I think your suggestion makes sense.
One edge case I can think of is the case where the resolver also sends address weights.
Then [A=2,B=1,A=1] would have to become [A=3, B=1] and so on. Possibly the resolver should be responsible for sending a proper resolution result so that we don't need implicit aggregation in the LB policy like this.

So my suggestion is to still have one counter per subchannel (or subchannel ref in C-core) and use exactly the same logic as round robin for deciding how duplicate addresses should translate into duplicate subchannels (or subchannel refs).

Reply:
@ejona86 tells me that round_robin in Java *does* de-dup subchannels for duplicate addresses (he pointed me at [this](https://github.com/grpc/grpc-java/blob/a2398ce5dbe2eacf841d7af9c9aa5e912e8db0aa/core/src/main/java/io/grpc/util/RoundRobinLoadBalancer.java#L82)), so there's only one connection created, but it still treats them as doubly weighted in terms of how it distributes RPCs across the addresses.  I think it would be consistent to do the same thing here.

For now, I think we can simply ignore duplicate addresses, since we're explicitly not supporting differently weighted endpoints.  But if we do add that later, we will want to treat this case as differently weighted endpoints.  (If we don't de-dup now, then it could cause behavior changes if we want to add support for differently weighted endpoints later.)

Given the different ways that address lists can be generated (DNS and EDS being two good examples), I think it makes sense for LB policies to understand both a weight attribute and the fact that a given address may be present in the list multiple times.  This is what we do in the ring_hash policy in C-core.

Reply:
Ahh, I see now, the EAGs in grpc-java have the same hashCode if the underlying addresses are equal. Indeed that would de-dup then :sweat_smile: 

> I think it would be consistent to do the same thing here.

@markdroth How would you like to phrase this in the proposal?
It's still not entirely clear to me if round_robin currently is consistent across languages in terms of handling duplicate addresses.

Maybe change the section further down to something like:

```
##### Duplicate addresses in resolution result

The `least_request_experimental` policy will be consistent with the `round_robin` policy in terms of handling duplicate addresses in the resolution result.

If support for the weighted least request algorithm is later added to the `least_request_experimental` policy, addresses should be de-duplicated and weighted accordingly just as the `ring_hash` policy does.
```

Reply:
It looks like all three languages de-dup to avoid creating redundant connections to the same address, but the weighting behavior is a little different.  Java and C-core both use it as a weighting indication, so that an address that is present twice in the list will get twice as many RPCs per pass through the list as an address that is present only once.  Go apparently does not currently do that weighting (i.e., an address that is present twice in the list will get just as many RPCs as an address that is present once), but @dfawley tells me he agrees that that's a bug, and he's noted it in grpc/grpc-go#2580 as something that needs to be fixed.  So I think we're generally in agreement about what the intended behavior is here.

In terms of the wording here, how about something like the following:

"""
In gRPC, if an address is present more than once in an address list, that is intended to indicate that the address should be more heavily weighted.  However, it should not result in more than one connection to that address.

In keeping with this, the `least_request_experimental` policy will de-duplicate the addresses in the address list it receives so that it does not create more than one connection to a given address.  However, because the policy will not initially support the weighted least request algorithm, we will ignore the additional weight given to the address.  If at some point in the future we do add support for the weighted least request algorithm, then an address that is present multiple times in the list will be given appropriate weight in that algorithm.
"""

Reply:
Sound good!
Updated now.